src/wp-admin/includes/image-edit.php

```diff
@@ -1037,10 +1139,233 @@ function wp_save_image( $post_id ) {
+ * @param array $subsized_images       Path of all subsized image file of all mime types.
+ * @return array Metadata with sources added.
+ */
+function wp_uploads_update_sources( $metadata, $valid_mime_transforms, $main_images, $subsized_images ) {
+	foreach ( $valid_mime_transforms as $targeted_mime ) {
+		// Make sure the path and file exists as those values are required.
+		$image_directory = null;
+		if ( isset( $main_images[ $targeted_mime ]['path'], $main_images[ $targeted_mime ]['file'] ) && file_exists( $main_images[ $targeted_mime ]['path'] ) ) {
+			// Add sources to original image metadata.
+			$metadata['sources'][ $targeted_mime ] = array(
+					'file'     => $main_images[ $targeted_mime ]['file'],
+					'filesize' => wp_filesize( $main_images[ $targeted_mime ]['path'] ),
+			);
+			$image_directory                       = pathinfo( $main_images[ $targeted_mime ]['path'], PATHINFO_DIRNAME );
+		}
+
+		/**
+		 * If no original image was provided the image_directory can't be determined, in that scenario try to
+		 * find it from the `file` property.
+		 *
+		 * @see get_attached_file()
+		 */
+		if (
+				null === $image_directory
+				&& isset( $metadata['file'] )
+				&& 0 !== strpos( $metadata['file'], '/' )
+				&& ':\\' !== substr( $metadata['file'], 1, 2 )
+		) {
+			$uploads = wp_get_upload_dir();
+			if ( false === $uploads['error'] && isset( $uploads['basedir'] ) ) {
+				$file = path_join( $uploads['basedir'], $metadata['file'] );
+				if ( file_exists( $file ) ) {
+					$image_directory = pathinfo( $file, PATHINFO_DIRNAME );
+				}
+			}
+		}
+
+		if ( null === $image_directory ) {
+			continue;
+		}
+
+		foreach ( $metadata['sizes'] as $size_name => $size_details ) {
+			if ( empty( $subsized_images[ $targeted_mime ][ $size_name ]['file'] ) ) {
+				continue;
+			}
+
+			// Add sources to resized image metadata.
+			$subsize_path = path_join( $image_directory, $subsized_images[ $targeted_mime ][ $size_name ]['file'] );
+			if ( ! file_exists( $subsize_path ) ) {
+				continue;
+			}
+
+			$metadata['sizes'][ $size_name ]['sources'][ $targeted_mime ] = array(
+					'file'     => $subsized_images[ $targeted_mime ][ $size_name ]['file'],
+					'filesize' => wp_filesize( $subsize_path ),
+			);
+		}
+	}
+
+	return $metadata;
+}
+
+/**
+ * Before saving the metadata of the image store a backup values for the sources and file property
+ * those files would be used and deleted by the backup mechanism, right after the metadata has
+ * been updated. It removes the current sources property due once this function is executed
+ * right after an edit has taken place and the current sources are no longer accurate.
+ *
+ * @since 6.1.0
+ *
+ * @param int   $attachment_id The ID representing the attachment.
+ * @param array $data          The current metadata of the attachment.
+ * @return array The updated metadata for the attachment.
+ */
+function wp_uploads_backup_sources( $attachment_id, $data ) {
```

Comment:
See above. This function should not be needed as such, we need to integrate the logic directly with core's logic. Specifically this function includes a bunch of hacks, adding actions temporarily etc., which indicates the hacky nature of the plugin implementation.